src/status_im/contexts/wallet/wallet_connect/utils/typed_data.cljs

```diff
@@ -0,0 +1,94 @@
+(ns status-im.contexts.wallet.wallet-connect.utils.typed-data
+  (:require [clojure.string :as string]
+            [status-im.contexts.wallet.wallet-connect.utils.networks :as networks]
+            [status-im.contexts.wallet.wallet-connect.utils.rpc :as rpc]
+            [utils.number :as number]))
+
+(declare flatten-data)
+
+(defn- format-flattened-key
+  [k]
+  (cond
+    (keyword? k) (name k)
+    (number? k)  (str k)
+    (string? k)  k
+    :else        "unsupported-key"))
+
+(defn- flatten-map
+  [data path]
+  (reduce-kv (fn [acc k v]
+               (->> (format-flattened-key k)
+                    (conj path)
+                    (flatten-data v acc)))
+             []
+             data))
+
+(defn- flatten-vec
+  [data path]
+  (->> data
+       (map-indexed vector)
+       (reduce (fn [acc [idx v]]
+                 (->> (str idx)
+                      (conj path)
+                      (flatten-data v acc)))
+               [])))
+
+(defn flatten-data
+  "Recursively flatten a map or vector into a flat vector.
+
+  e.g. `[[[\"person\" \"first-name\"] \"Rich\"]
+        [[[\"person\" \"last-name\"] \"Hickey\"]]]`"
+  ([value]
+   (flatten-data value [] []))
+  ([value acc path]
+   (cond
+     (map? value)    (into acc (flatten-map value path))
+     (vector? value) (into acc (flatten-vec value path))
+     :else           (conj acc [path value]))))
+
+(defn format-fields
+  "Format the fields into maps with `:label` & `:value`, where the label
+  is the flattened keys joined with a separator
+
+  e.g. `{:label \"person: first-name:\" :value \"Rich\"}`"
+  [data separator]
+  (mapv (fn [[kv v]]
+          {:label (-> separator
+                      (string/join kv)
+                      (str separator)
+                      string/trim)
+           :value v})
+        data))
+
+(defn flatten-typed-data
+  "Flatten typed data and prepare it for UI"
+  [typed-data]
+  (-> typed-data
+      (select-keys [:domain :message])
+      flatten-data
+      (format-fields ": ")))
```

Comment:
Is the format `from: wallets: 1:` copied from other products or is this a standard we are creating ourselves for Status? I find hard to imagine users can figure out the indexes and the flattening going on, but maybe not, things are crazy in terms of usability when signing.

Do you know if there's any interest in improving the designs or if any concerns were raised?